core-backend/sandbox/src/funcs.rs

```diff
@@ -122,188 +126,304 @@ where
     pub fn send(ctx: &mut Runtime<E>, args: &[Value]) -> SyscallOutput {
         sys_trace!(target: "syscall::gear", "send, args = {}", args_to_str(args));
 
-        let (destination_ptr, payload_ptr, payload_len, value_ptr, delay, message_id_ptr) =
-            args.iter().read_6()?;
+        let (pid_value_ptr, payload_ptr, len, delay, err_mid_ptr) = args.iter().read_5()?;
 
         ctx.run(|ctx| {
-            let destination = ctx.read_memory_as(destination_ptr)?;
-            let payload = ctx.read_memory(payload_ptr, payload_len)?.try_into()?;
-            let value = ctx.read_memory_as(value_ptr)?;
+            let HashWithValue {
+                hash: destination,
+                value,
+            } = ctx.read_memory_as(pid_value_ptr)?;
+            let payload = ctx.read_memory(payload_ptr, len)?.try_into()?;
 
             ctx.ext
-                .send(HandlePacket::new(destination, payload, value), delay)
+                .send(HandlePacket::new(destination.into(), payload, value), delay)
                 .process_error()
                 .map_err(FuncError::Core)?
                 .error_len_on_success(|message_id| {
-                    ctx.write_output(message_id_ptr, message_id.as_ref())
+                    ctx.write_memory_as(
+                        err_mid_ptr,
+                        LengthWithHash {
+                            hash: message_id.into(),
+                            ..Default::default()
+                        },
+                    )
+                })
+                .and_then(|length| {
```

Comment:
what if `error_len_on_success` returns Ok(0) ? You will overwrite `message_id` ? Also we should guaranty that error encoded size (when it is really error case) is not zero.